### PR TITLE
Input option child access

### LINKED ISSUE #668: Support input option child access
the following line: 
```Input<Option<FilteredGameState>, "filtered_game_controller_state?.game_state">```
produces the following error:
```error[E0308]: mismatched types
    --> /home/josi/worktree/hulk/target/incremental/build/hulk-ec3e57a7c7babecf/out/generated_code.rs:3110:45
     |
3093 |   ...                   .cycle(control::world_state_composer::CycleContext::new(
     |                                ------------------------------------------------ arguments to this function are incorrect
...
3110 | / ...                       (|| {
3111 | | ...                           Some(
3112 | | ...                               own_database_reference
3113 | | ...                                   .main_outputs
...    |
3117 | | ...                           )
3118 | | ...                       })(),
     | |______________________________^ expected `Option<&FilteredGameState>`, found `Option<FilteredGameState>`
     |
     = note: expected enum `std::option::Option<&FilteredGameState>`
                found enum `std::option::Option<FilteredGameState>`
note: associated function defined here
    --> /home/josi/worktree/hulk/crates/control/src/world_state_composer.rs:25:1
     |
25   | #[context]
     | ^^^^^^^^^^
     = note: this error originates in the attribute macro `context` (in Nightly builds, run with -Z macro-backtrace for more info)
help: try using `.as_ref()` to convert `std::option::Option<FilteredGameState>` to `std::option::Option<&FilteredGameState>`
     |
3118 |                                             })().as_ref(),
     |                                                 +++++++++
```

### PR DESCRIPTION
**Pull request by ichbinJosi**
## Introduced Changes

Fixed ```path_to_accessor_token_stream_with_cycler_instance```

Fixes #668

### CODE REVIEW

**Comment by schmidma** [comment]
This seems to be based on #666 . Is this necessary? Can you rebase this PR onto main?

**Comment by schluis** [comment]
Yep, rebased onto main

File: crates/code_generation/src/accessor.rs [hunk 1]
```diff
@@ -55,28 +56,52 @@ fn path_to_accessor_token_stream_with_cycler_instance(
...
+        ReferenceKind::Mutable => quote! {&mut },
+    };
+
     if path.contains_optional() {
-        quote! {
-            (|| Some(#prefix . #(#segments).*))()
-        }
+        let first_segment = segments
+            .next()
+            .expect("Path should always contain at least one segment");
```
[the remark this answers is not in the document]
Comment (ichbinJosi) [reply crates/code_generation/src/accessor.rs:82]: Done

File: crates/code_generation/src/accessor.rs [hunk 2]
```diff
@@ -55,28 +56,52 @@ fn path_to_accessor_token_stream_with_cycler_instance(
...
+                }
+                EitherOrBoth::Right(previous_segment) => match previous_segment.is_optional {
+                    true => quote! {
+                        (|| Some(#reference (*#recursive_token_stream)))()
+                    },
+                    false => quote! {
+                        (|| Some(#recursive_token_stream))()
+                    },
+                },
+                EitherOrBoth::Left(_) => unimplemented!(),
```
[the remark this answers is not in the document]
Comment (ichbinJosi) [reply crates/code_generation/src/accessor.rs:101]: Done

File: crates/code_generation/src/accessor.rs [hunk 3]
```diff
@@ -55,28 +56,52 @@ fn path_to_accessor_token_stream_with_cycler_instance(
...
     if path.contains_optional() {
-        quote! {
-            (|| Some(#prefix . #(#segments).*))()
-        }
+        let first_segment = segments
+            .next()
+            .expect("Path should always contain at least one segment");
+
+        segments.zip_longest(path.segments.iter()).fold(
+            quote! {#prefix. #first_segment},
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/code_generation/src/accessor.rs:85]: @okiwi6 and @schluis decided that this is okay. I'll resolve as commanded :vulcan_salute:

**Review by schmidma — APPROVED** [review]

**Pull request merged**